File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    
    if ( !(head) || !(head.next) ) return null;
    
    let fp = head;
    let bp = head;
    
    while (n--) {
        bp = bp.next;
    }
    
    // Remove first element
    if ( !bp ) return head.next;        
        
    while (bp.next) {
        fp = fp.next;
        bp = bp.next;
    }
    
    let saveNext = fp.next.next;
    delete fp.next;
    fp.next = saveNext;
    
    return head;
};

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        if ( !(head -> next) ) return nullptr;
        
        int check = _removeNode(head, n); 
        
        // if check == 0 => 1st item needs to be removed.
        
        if (check == 0) head = head -> next;
        
        return head;
    }
    
    int _removeNode(ListNode* node, int n) {
       
        if (node == nullptr) return n;
        
        int jumpForNext = _removeNode(node -> next, n);
        
        if (jumpForNext == 0) {
            // found
            ListNode* saveNext = node -> next -> next;
            delete node -> next;
            node -> next = saveNext;
        }
        
        return jumpForNext - 1;
    }
};